src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,178 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private transient MyLinkedList.Node<T> first;
+    private transient MyLinkedList.Node<T> last;
+    private transient int size = 0;
+
     @Override
     public void add(T value) {
+        Node<T> val = new Node<>(null, value, null);
+        if (size == 0) {
+            first = last = val;
+        } else {
+            swapLast(val);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> givenNode = new Node<>(null, value, null);
+        Node<T> current = node(index);
+
+        if (index == 0) {
+            swapFirst(givenNode);
+        } else if (index == size) {
+            swapLast(givenNode);
+        } else {
+            current.prev.next = givenNode;
+            givenNode.prev = current.prev;
+
+            current.prev = givenNode;
+            givenNode.next = current;
+        }
+        size++;
+    }
+
+    private void swapFirst(Node<T> node) {
+        if (size == 0) {
+            first = last = node;
+            return;
+        }
+        first.prev = node;
+        node.next = first;
+        first = node;
+    }
+
+    private void swapLast(Node<T> node) {
+        if (size == 0) {
+            first = last = node;
+            return;
+        }
+        last.next = node;
+        node.prev = last;
+        last = node;
+    }
+
+    MyLinkedList.Node<T> node(int index) {
+        Node<T> x;
+        if (index < (size >> 1)) {
+            x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+        } else {
+            x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+        }
+        return x;
+    }
+
+    private void checkElementIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list.isEmpty()) {
+            return;
+        }
+        for (T node : list) {
+            add(node);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        newNode.item = node(index).item;
+        node(index).item = value;
+        return newNode.item;
```

Comment:
The `set` method should not create a new node. Instead, it should update the `item` of the existing node at the specified index. Creating a new node is unnecessary and inefficient.